src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -145,28 +140,28 @@ public override IEnumerable CreateDerivedEnumerableInstance(ref ReadStack state,
                     return instanceOfIList;
                 }
             }
-            else if (instance is ICollection<TRuntimeProperty> instanceOfICollection)
+            else if (instance is ICollection<TDeclaredProperty> instanceOfICollection)
```

Comment:
It's not a rename - in some cases, we were previously casting strongly typed collection elements to their runtime type. This doesn't work for nullable types because in some cases, the runtime type is the non-nullable version of that type (e.g. `float?` (declared type) -> `float` (runtime type)).

We can't cast the parent collection to be returned in such cases (e.g. List<float>to List<float?>), so we should be using the declared type.